pkg/datapath/tables/node_address.go

```diff
@@ -0,0 +1,365 @@
+	NodeAddresses   statedb.RWTable[NodeAddress]
+	AddressScopeMax AddressScopeMax
+}
+
+type nodeAddressController struct {
+	nodeAddressControllerParams
+
+	tracker *statedb.DeleteTracker[*Device]
+}
+
+// newNodeAddressTable constructs the node address controller & registers its
+// lifecycle hooks and then provides Table[NodeAddress] to the application.
+// This enforces proper ordering, e.g. controller is started before anything
+// that depends on Table[NodeAddress] and allows it to populate it before
+// it is accessed.
+func newNodeAddressTable(p nodeAddressControllerParams) (tbl statedb.Table[NodeAddress], err error) {
+	n := nodeAddressController{nodeAddressControllerParams: p}
+	n.register()
+	return n.NodeAddresses, nil
+}
+
+func (n *nodeAddressController) register() {
+	g := n.Jobs.NewGroup(n.HealthScope)
+	g.Add(job.OneShot("node-address-update", n.run))
+
+	n.Lifecycle.Append(
+		hive.Hook{
+			OnStart: func(ctx hive.HookContext) error {
+				txn := n.DB.WriteTxn(n.NodeAddresses, n.Devices /* for delete tracker */)
+				defer txn.Abort()
+
+				// Start tracking deletions of devices.
+				var err error
+				n.tracker, err = n.Devices.DeleteTracker(txn, "node-addresses")
+				if err != nil {
+					return fmt.Errorf("DeleteTracker: %w", err)
+				}
+
+				// Do an immediate update to populate the table before it is read from.
+				devices, _ := n.Devices.All(txn)
+				for dev, _, ok := devices.Next(); ok; dev, _, ok = devices.Next() {
+					n.update(txn, nil, n.getAddressesFromDevice(dev), nil)
+				}
+				txn.Commit()
+
+				// Start the job in the background to incremental refresh
+				// the node addresses.
+				return g.Start(ctx)
+			},
+			OnStop: g.Stop,
+		})
+
+}
+
+func (n *nodeAddressController) run(ctx context.Context, reporter cell.HealthReporter) error {
+	defer n.tracker.Close()
+
+	limiter := rate.NewLimiter(nodeAddressControllerMinInterval, 1)
+	revision := statedb.Revision(0)
+	for {
+		txn := n.DB.WriteTxn(n.NodeAddresses)
+		process := func(dev *Device, deleted bool, rev statedb.Revision) error {
+			addrIter, _ := n.NodeAddresses.Get(txn, NodeAddressDeviceNameIndex.Query(dev.Name))
+			existing := sets.New[NodeAddress](statedb.Collect[NodeAddress](addrIter)...)
+			var new sets.Set[NodeAddress]
+			if !deleted {
+				new = n.getAddressesFromDevice(dev)
+			}
+			n.update(txn, existing, new, reporter)
+			return nil
+		}
+		var watch <-chan struct{}
+		revision, watch, _ = n.tracker.Process(txn, revision, process)
+		txn.Commit()
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-watch:
+		}
+		if err := limiter.Wait(ctx); err != nil {
+			return err
+		}
+	}
+}
+
+func (n *nodeAddressController) update(txn statedb.WriteTxn, existing, new sets.Set[NodeAddress], reporter cell.HealthReporter) {
+	updated := false
+
+	// Insert new addresses that did not exist.
+	for addr := range new {
+		if !existing.Has(addr) {
+			updated = true
+			n.NodeAddresses.Insert(txn, addr)
```

Comment:
can we `break` after the insertion here, given that `addr` should be the primary key?

Reply:
No because we need to add all new addresses. Were just checking first if the address already exists to avoid mutation of the table and thus unnecessary triggering of watchers.